src/common/acl_principal.c

```diff
@@ -0,0 +1,344 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+ * The Government's rights to use, modify, reproduce, release, perform, display,
+ * or disclose this software are subject to the terms of the Apache License as
+ * provided in Contract No. 8F-30005.
+ * Any reproduction of computer software, computer software documentation, or
+ * portions thereof marked with this legend must also reproduce the markings.
+ */
+/*
+ * daos_acl: This file contains functions related to working with the principals
+ * from Access Control Lists.
+ */
+
+#include <daos_security.h>
+#include <gurt/common.h>
+#include <pwd.h>
+#include <grp.h>
+
+#define DEFAULT_BUF_LEN 1024
+
+/*
+ * States used to parse a principal name
+ */
+enum validity_state {
+	STATE_START,
+	STATE_NAME,
+	STATE_DOMAIN,
+	STATE_DONE,
+	STATE_INVALID
+};
+
+static enum validity_state
+next_validity_state(enum validity_state current_state, char ch)
+{
+	switch (current_state) {
+	case STATE_START:
+		if (ch == '@')
+			return STATE_INVALID;
+		return STATE_NAME;
+	case STATE_NAME:
+		if (ch == '@')
+			return STATE_DOMAIN;
+		if (ch == '\0')
+			return STATE_INVALID;
+		break;
+	case STATE_DOMAIN:
+		if (ch == '\0')
+			return STATE_DONE;
+		if (ch == '@')
+			return STATE_INVALID;
+		break;
+	case STATE_DONE:
+	case STATE_INVALID:
+		break;
+	default:
+		D_ASSERTF(false, "Invalid state: %d\n", current_state);
+	}
+
+	return current_state;
+}
+
+bool
+daos_acl_principal_is_valid(const char *name)
+{
+	size_t			len;
+	size_t			i;
+	enum validity_state	state = STATE_START;
+
+	if (name == NULL) {
+		D_INFO("Name was NULL\n");
+		return false;
+	}
+
+	len = strnlen(name, DAOS_ACL_MAX_PRINCIPAL_BUF_LEN);
+	if (len == 0 || len > DAOS_ACL_MAX_PRINCIPAL_LEN) {
+		D_INFO("Invalid len: %lu\n", len);
+		return false;
+	}
+
+	for (i = 0; i < (len + 1); i++) {
+		state = next_validity_state(state, name[i]);
+		if (state == STATE_INVALID) {
+			D_INFO("Name was badly formatted: %s\n", name);
+			return false;
+		}
+	}
+
+	return true;
+}
+
+static int
+local_name_to_principal_name(const char *local_name, char **name)
+{
+	D_ASPRINTF(*name, "%s@", local_name);
+	if (*name == NULL) {
+		D_ERROR("Failed to allocate string for name\n");
+		return -DER_NOMEM;
+	}
+
+	return 0;
+}
+
+int
+daos_acl_uid_to_principal(uid_t uid, char **name)
+{
+	int		rc;
+	struct passwd	user;
+	struct passwd	*result = NULL;
+	char		*buf = NULL;
+	char		*new_buf = NULL;
+	size_t		buflen = DEFAULT_BUF_LEN;
+
+	if (name == NULL) {
+		D_INFO("name pointer was NULL!\n");
+		return -DER_INVAL;
+	}
+
+	/*
+	 * No platform-agnostic way to fetch the max buflen - so let's try a
+	 * sane value and double until it's big enough.
+	 */
+	do {
+		D_REALLOC(new_buf, buf, buflen);
+		if (new_buf == NULL) {
+			D_ERROR("Couldn't allocate memory for getpwuid_r\n");
+			return -DER_NOMEM;
```

Comment:
There is a memory leak here.   If realloc fails, the original block is left untouched so one time around the loop would mean buf is allocated.   You can do D_GOTO(out, rc = -DER_NOMEM); here since D_FREE can safely be called on a NULL pointer to cover the first time around the loop.

Reply:
Good catch.